src/java.base/share/classes/sun/security/ssl/X509KeyManagerImpl.java

```diff
@@ -180,6 +301,116 @@ public String[] getServerAliases(String keyType, Principal[] issuers) {
     // implementation private methods
     //
 
+    private String removeAliasIndex (String alias){
+        if (alias == null) {
+            return null;
+        }
+
+        int firstDot = alias.indexOf('.');
+        int secondDot = alias.indexOf('.', firstDot + 1);
+        if ((firstDot == -1) || (secondDot == firstDot)) {
+            // invalid alias
+            return null;
+        }
+        return alias.substring(firstDot + 1);
+    }
+
+    private boolean areCertChainsEqual(String builderAlias,
+            X509Certificate[] cachedCerts) {
+        int aDot = builderAlias.indexOf('.');
+        int builderIndex = Integer.parseInt(builderAlias.substring(0, aDot));
+        String entryAlias = builderAlias.substring(aDot + 1);
+        try {
+            Builder builder = builders.get(builderIndex);
+            KeyStore keyStore = builder.getKeyStore();
+            if (keyStore == null) {
+                return false;
+            }
+            Certificate[] keyStoreCerts = keyStore.getCertificateChain(entryAlias);
+            if (keyStoreCerts == null || keyStoreCerts.length == 0) {
+                return false;
+            }
+
+            // Convert KeyStore certificates to X509Certificates
+            X509Certificate[] newCerts = Arrays.copyOf(keyStoreCerts,
```

Comment:
Does it have to convert `Certificate[]` to `X509Certificate[]`?
The doc for `Arrays::equals` states
> the two arrays are equal if they contain the same elements in the same order.

It looks this method doesn't care the types of the arrays themselves.

Reply:
Removed conversion.